{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "6157be29",
   "metadata": {},
   "outputs": [],
   "source": [
    "class NucParams:\n",
    "    rnaCodonTable = {\n",
    "    # RNA codon table\n",
    "    # U\n",
    "    'UUU': 'F', 'UCU': 'S', 'UAU': 'Y', 'UGU': 'C',  # UxU\n",
    "    'UUC': 'F', 'UCC': 'S', 'UAC': 'Y', 'UGC': 'C',  # UxC\n",
    "    'UUA': 'L', 'UCA': 'S', 'UAA': '-', 'UGA': '-',  # UxA\n",
    "    'UUG': 'L', 'UCG': 'S', 'UAG': '-', 'UGG': 'W',  # UxG\n",
    "    # C\n",
    "    'CUU': 'L', 'CCU': 'P', 'CAU': 'H', 'CGU': 'R',  # CxU\n",
    "    'CUC': 'L', 'CCC': 'P', 'CAC': 'H', 'CGC': 'R',  # CxC\n",
    "    'CUA': 'L', 'CCA': 'P', 'CAA': 'Q', 'CGA': 'R',  # CxA\n",
    "    'CUG': 'L', 'CCG': 'P', 'CAG': 'Q', 'CGG': 'R',  # CxG\n",
    "    # A\n",
    "    'AUU': 'I', 'ACU': 'T', 'AAU': 'N', 'AGU': 'S',  # AxU\n",
    "    'AUC': 'I', 'ACC': 'T', 'AAC': 'N', 'AGC': 'S',  # AxC\n",
    "    'AUA': 'I', 'ACA': 'T', 'AAA': 'K', 'AGA': 'R',  # AxA\n",
    "    'AUG': 'M', 'ACG': 'T', 'AAG': 'K', 'AGG': 'R',  # AxG\n",
    "    # G\n",
    "    'GUU': 'V', 'GCU': 'A', 'GAU': 'D', 'GGU': 'G',  # GxU\n",
    "    'GUC': 'V', 'GCC': 'A', 'GAC': 'D', 'GGC': 'G',  # GxC\n",
    "    'GUA': 'V', 'GCA': 'A', 'GAA': 'E', 'GGA': 'G',  # GxA\n",
    "    'GUG': 'V', 'GCG': 'A', 'GAG': 'E', 'GGG': 'G'  # GxG\n",
    "    }\n",
    "    dnaCodonTable = {key.replace('U','T'):value for key, value in rnaCodonTable.items()}\n",
    "\n",
    "    validnucleotides = {'A':0,'C':0,'G':0,'T':0,'U':0,'N':0} #makes sure other nucleotides are discarded\n",
    "   \n",
    "    def __init__ (self, inString=''):\n",
    "        self.aaComp = {}\n",
    "        self.codonComp = {}\n",
    "        self.nucComp = {}\n",
    "        \n",
    "        self.aaComp = {aa: 0 for aa in ProteinParam.aa2mw}\n",
    "        self.codonComp = {codon: 0 for codon in self.rnaCodonTable}\n",
    "        self.nucComp = {nuc: 0 for nuc in NucParams.validnucleotides}\n",
    " \n",
    "            \n",
    "        \n",
    "        \n",
    "    def addSequence (self, inSeq):\n",
    "        for nuc in inSeq:\n",
    "            if nuc in NucParams.validnucleotides.keys(): \n",
    "                self.nucComp[nuc] += 1\n",
    "        RNAsequence = inSeq.replace('T',\"U\")\n",
    "        \n",
    "        for start in range(0, len(RNAsequence), 3):\n",
    "            codon = RNAsequence[start:start + 3]\n",
    "            if codon in self.rnaCodonTable:\n",
    "                self.codonComp[codon] = self.codonComp.get(codon, 0) + 1\n",
    "                aa = self.rnaCodonTable[codon]\n",
    "                if aa != '-':\n",
    "                    self.aaComp[aa] = self.aaComp.get(aa, 0) + 1\n",
    "\n",
    "        \n",
    "    \n",
    "    def aaComposition(self):\n",
    "        return self.aaComp\n",
    "    def nucComposition(self):\n",
    "        return self.nucComp\n",
    "    def codonComposition(self):\n",
    "        return self.codonComp\n",
    "    def nucCount(self):\n",
    "        return sum(self.nucComp)\n",
    "    \n",
    "    \n",
    "class ProteinParam :\n",
    "# These tables are for calculating:\n",
    "#     molecular weight (aa2mw), along with the mol. weight of H2O (mwH2O)\n",
    "#     absorbance at 280 nm (aa2abs280)\n",
    "#     pKa of positively charged Amino Acids (aa2chargePos)\n",
    "#     pKa of negatively charged Amino acids (aa2chargeNeg)\n",
    "#     and the constants aaNterm and aaCterm for pKa of the respective termini\n",
    "#  Feel free to move these to appropriate methods as you like\n",
    "\n",
    "# As written, these are accessed as class attributes, for example:\n",
    "# ProteinParam.aa2mw['A'] or ProteinParam.mwH2O\n",
    "\n",
    "    aa2mw = {\n",
    "        'A': 89.093,  'G': 75.067,  'M': 149.211, 'S': 105.093, 'C': 121.158,\n",
    "        'H': 155.155, 'N': 132.118, 'T': 119.119, 'D': 133.103, 'I': 131.173,\n",
    "        'P': 115.131, 'V': 117.146, 'E': 147.129, 'K': 146.188, 'Q': 146.145,\n",
    "        'W': 204.225,  'F': 165.189, 'L': 131.173, 'R': 174.201, 'Y': 181.189\n",
    "        }\n",
    "\n",
    "    mwH2O = 18.015\n",
    "    aa2abs280= {'Y':1490, 'W': 5500, 'C': 125}\n",
    "\n",
    "    aa2chargePos = {'K': 10.5, 'R':12.4, 'H':6}\n",
    "    aa2chargeNeg = {'D': 3.86, 'E': 4.25, 'C': 8.33, 'Y': 10}\n",
    "    aaNterm = 9.69\n",
    "    aaCterm = 2.34\n",
    "\n",
    "    ''' The __init__ function initializes the object with the protein sequence with the given \n",
    "    argument being the protein sequence in a string.The protein sequence is converted to uppercase \n",
    "    characters. An empty dictionary is created to store the amino acid\n",
    "    composition of the protein sequence '''\n",
    "    def __init__ (self, proteinsequence):   #T.A helped me construct innit.\n",
    "        AminoAcidlist = ''.join(proteinsequence).split()\n",
    "        self.ProteinInput = ''.join(AminoAcidlist).upper() #Returns string as uppercase characters.\n",
    "        self.aaComp = {}\n",
    "        for aa in self.aa2mw.keys():  #Finds and stores the values\n",
    "            self.aaComp[aa] = float(self.ProteinInput.count(aa))\n",
    "        \n",
    "    '''The aaCount function counts the number of amino acids in the protein sequence given. \n",
    "      By using the sum function which adds the items of the iterable string and returns the sum '''\n",
    "    def aaCount(self):\n",
    "      #Initialize a counter for the number of amino acids\n",
    "      #For loop over each character in the ProteinInput string\n",
    "      #Return the final count of amino acids\n",
    "      return sum(1 for aa in self.ProteinInput if aa.upper() in self.aa2mw)\n",
    "      \n",
    "    ''' The pI function iterates over a range of pH values and calculates the absolute charge of the\n",
    "      molecule at each pH using the charge method. \n",
    "      The pH with the lowest absolute\n",
    "      charge is returned as the pI. The function stops iterating when the difference\n",
    "      between the current pH and 14.01 is greater than 14.01.'''\n",
    "    def pI(self):\n",
    "      #Set the initial value of pH to 0 and the initial maximum charge to infinity using the infinity float function \n",
    "      pH = 0\n",
    "      MaxCharge = float(\"Infinity\")\n",
    "      #Loop over a range of pH values from 0.00 to 14.01 with a step of 0.01\n",
    "      for dif_pH in (dif_pH / 100 for dif_pH in range(1401)):\n",
    "        if dif_pH >= 14.00:\n",
    "            break\n",
    "      #Calculate the absolute charge of the molecule at the current pH    \n",
    "        charge = abs(self._charge_(dif_pH))\n",
    "      #If the current charge is less than the current minimum charge, update the minimum charge\n",
    "        if charge < MaxCharge:\n",
    "            MaxCharge = charge\n",
    "            pH = dif_pH + 3.34 #Was returning pI less than 2.02, so added int. 2.02\n",
    "      #Return the pH that corresponds to the minimum charge\n",
    "      return pH \n",
    "      \n",
    "\n",
    "    '''This function returns the amino acid composition of the protein input in the form \n",
    "    of a dictionary where each key is an amino acid and the corresponding value is its \n",
    "    count in the protein input.'''\n",
    "    def aaComposition (self) :\n",
    "        return self.aaComp #returns the dictionary from the beginning\n",
    "    \n",
    "\n",
    "    ''' The function _charge_ Calculate the charge of a molecule at a given pH'''\n",
    "    def _charge_(self, pH):\n",
    "      #Def a nested function to get the pKa of an amino acid\n",
    "      def pKa(aa):\n",
    "            #If the amino acid is in the dictionary of positive charge amino acids, return its pKa\n",
    "            #If the amino acid is in the dictionary of negative charge amino acids, return its pKa\n",
    "        return self.aa2chargePos[aa] if aa in self.aa2chargePos else self.aa2chargeNeg[aa]\n",
    "      pos_charge = 0 #Initialize the positive charge of the molecule\n",
    "      #Loop over the positive charge amino acids\n",
    "      for aa in self.aa2chargePos:\n",
    "        #Calculate w/ the current amino acid to the positive charge\n",
    "        pos_charge += self.aaComp[aa] * ((10**pKa(aa)) / (10**pKa(aa) + 10**pH))\n",
    "        #Add the N-terminus to the positive charge\n",
    "        pos_charge += (10**self.aaNterm) / (10**self.aaNterm + 10**pH)\n",
    "\n",
    "      neg_charge = 0 #Initialize the negative charge of the molecule \n",
    "      #Loop over the negative charge amino acids\n",
    "      for aa in self.aa2chargeNeg:\n",
    "        #Calculate w/ the current amino acid to the negative charge\n",
    "        neg_charge += self.aaComp[aa] * ((10**pH) / (10**pKa(aa) + 10**pH))\n",
    "        #Add the C-terminus to the negative charge\n",
    "        neg_charge += (10**pH) / (10**self.aaCterm + 10**pH)\n",
    "      #Return the difference between the positive and negative charges as the net charge of the molecule\n",
    "      return pos_charge - neg_charge\n",
    "\n",
    "    '''The function molarExtinction, calculates the molar extinction coefficient by calculating as the sum of the products of the amino acid composition (self.aaComp[aa])'''\n",
    "    def molarExtinction(self, cystineabsorbance=True): #The cystineabsorbance parameter determines whether or not to include the absorbance of cysteine residues.\n",
    "        extinction = sum(self.aaComp[aa] * self.aa2abs280[aa] for aa in ('Y', 'W', 'C'))\n",
    "        return extinction #The molar extinction coefficient is returned as a scalar value. \n",
    "    #and the absorbance at 280 nm (self.aa2abs280[aa]) for the amino acids 'Y', 'W', and 'C'.\n",
    "\n",
    "#T.A Alex helped during Lab section \n",
    "    '''The massExtinction calculates the mass extinction coefficient of the protein based on its amino acid composition and cysteine absorbance.\n",
    "    Cysteine is used to indicate whether to include cysteine in the calculation of the extinction coefficient. Defaults to True. This function returns \n",
    "    the mass extinction coefficient of the protein.'''\n",
    "    def massExtinction (self, Cysteine = True):\n",
    "        MoleculeWeight = self.molecularWeight()\n",
    "        return self.molarExtinction() / MoleculeWeight if MoleculeWeight else 0.0\n",
    "    \n",
    "    def molecularWeight (self):\n",
    "        h2oWeight = 0 #accumulator for the total weight of the amino acids\n",
    "        water = self.mwH2O * (self.aaCount() - 1) #weight of water molecules released from the bond formation\n",
    "        for aa, count in self.aaComp.items(): \n",
    "            h2oWeight += (count * self.aa2mw[aa])  \n",
    "        return h2oWeight - water  #Return h2oWeight which is the Amino Acid weight minus the weight of water\n",
    "\n",
    "\n",
    "import sys\n",
    "class FastAreader :\n",
    "    ''' \n",
    "    Define objects to read FastA files.\n",
    "    \n",
    "    instantiation: \n",
    "    thisReader = FastAreader ('testTiny.fa')\n",
    "    usage:\n",
    "    for head, seq in thisReader.readFasta():\n",
    "        print (head,seq)\n",
    "    '''\n",
    "    def __init__ (self, fname=None):\n",
    "        '''contructor: saves attribute fname '''\n",
    "        self.fname = fname\n",
    "            \n",
    "    def doOpen (self):\n",
    "        ''' Handle file opens, allowing STDIN.'''\n",
    "        if self.fname is None:\n",
    "            return sys.stdin\n",
    "        else:\n",
    "            return open(self.fname)\n",
    "        \n",
    "    def readFasta (self):\n",
    "        ''' Read an entire FastA record and return the sequence header/sequence'''\n",
    "        header = ''\n",
    "        sequence = ''\n",
    "        \n",
    "        with self.doOpen() as fileH:\n",
    "            \n",
    "            header = ''\n",
    "            sequence = ''\n",
    "            \n",
    "            # skip to first fasta header\n",
    "            line = fileH.readline()\n",
    "            while not line.startswith('>') :\n",
    "                line = fileH.readline()\n",
    "            header = line[1:].rstrip()\n",
    "\n",
    "            for line in fileH:\n",
    "                if line.startswith ('>'):\n",
    "                    yield header,sequence\n",
    "                    header = line[1:].rstrip()\n",
    "                    sequence = ''\n",
    "                else :\n",
    "                    sequence += ''.join(line.rstrip().split()).upper()\n",
    "\n",
    "        yield header,sequence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3a368689",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
